src/libutil/include/nix/util/url.hh

```diff
@@ -75,12 +81,141 @@ struct ParsedURL
      * part of the URL.
      */
     std::optional<Authority> authority;
-    std::string path;
+
+    /**
+     * @note Unlike Unix paths, URLs provide a way to escape path
+     * separators, in the form of the `%2F` encoding of `/`. That means
+     * that if one percent-decodes the path into a single string, that
+     * decoding will be *lossy*, because `/` and `%2F` both become `/`.
+     * The right thing to do is instead split up the path on `/`, and
+     * then percent decode each part.
+     *
+     * For an example, the path
+     * ```
+     * foo/bar%2Fbaz/quux
+     * ```
+     * is parsed as
+     * ```
+     * {"foo, "bar/baz", "quux"}
+     * ```
+     *
+     * Interior slashes are handled by the `std::vector` --- problem
+     * solved. Leading slashes are also not a problem, because they stem
+     * from the presence/absence of the authority (and a non-empty
+     * path). For examaple:
+     *
+     * ```
+     * tel:01234
+     * ```
+     * has path `{"01234"}` and no authority, while:
+     *
+     * ```
+     * file:///01234
+     * ```
+     * has path `{"", "01234"}` and an empty authority.
+     *
+     * Trailing slashes, however, are a problem.
+     * ```
+     * https://foo.com/bar
+     * https://foo.com/bar/
+     * ```
+     * are not the same, even after normalization. (One reason they
+     * cannot be normalized to the same thing is because relative URL
+     * parsing needs to treat them different, for the resolution of `..`
+     * path segments. See `parseURLRelative` for more details.)
+     *
+     * To handle this, and also redundant slashes which are normalized,
+     * the use of empty path segments is employed. For example, the two
+     * URLs above are parsed as follows:
+     *
+     * - ```
+     *   https://foo.com/bar
+     *   ```
+     *   has path
+     *   ```
+     *   {"", "bar"}
+     *   ```
+     *
+     * - ```
+     *   https://foo.com/bar/
+     *   ```
+     *   has path
+     *   ```
+     *   {"", "bar", ""}
+     *   ```
+     *
+     * - ```
+     *   https://foo.com//bar///
+     *   ```
+     *   has path
+     *   ```
+     *   {"", "", "bar", "", "", ""}
+     *   ```
```

Comment:
Now the same as how leading slashes are dealt with, so can talk about those together